addon_imps/storage/gitlab.py

```diff
@@ -0,0 +1,182 @@
+        await self.list_child_items(parsed_id.raw_id)
+        return ItemResult(
+            item_name=parsed_id.file_path.split("/")[-1],
+            item_id=parsed_id.raw_id,
+            item_type=ItemType.FOLDER,
+        )
+
+    async def _get_file(self, parsed_id, ref):
+        async with self.network.GET(
+            f"projects/{parsed_id.repo_id}/repository/files/{quote_plus(parsed_id.file_path)}",
+            query={"ref": ref},
+        ) as response:
+            content = await response.json_content()
+            if "file_name" not in content:
+                return None
+            return ItemResult(
+                item_name=content["file_name"],
+                item_id=parsed_id.raw_id,
+                item_type=ItemType.FILE,
+            )
+
+    async def list_child_items(
+        self,
+        item_id: str,
+        page_cursor: str = "",
+        item_type: storage.ItemType | None = None,
+    ) -> storage.ItemSampleResult:
+        parsed_id = ItemId.parse(item_id)
+        query_params = self._page_cursor_or_query(
+            page_cursor,
+            {
+                "pagination": "keyset",
+                "path": parsed_id.file_path,
+                "sort": "asc",
+                "order_by": "name",
+            },
+        )
+        async with self.network.GET(
+            f"projects/{parsed_id.repo_id}/repository/tree",
+            query=query_params,
+        ) as response:
+            if response.http_status == HTTPStatus.NOT_FOUND:
+                raise ItemNotFound
+            content = await response.json_content()
+            res_items = [parse_item(parsed_id.repo_id, item) for item in content]
+
+            if item_type:
+                res_items = [item for item in res_items if item.item_type == item_type]
+            return ItemSampleResult(
+                items=res_items,
+                next_sample_cursor=self._get_next_cursor(response.headers),
+            )
+
+
+@dataclass(frozen=True)
+class ItemId:
+    repo_id: str
+    file_path: str
+    raw_id: str
+
+    @classmethod
+    def parse(cls, item_id: str) -> ItemId:
+        repo_id, file_path = item_id.split(":", maxsplit=1)
+        return cls(
+            repo_id=repo_id,
+            file_path=file_path,
+            raw_id=item_id,
+        )
+
+
+def parse_item(repo_id: str, raw_item: dict) -> ItemResult:
+    return ItemResult(
+        item_id=f'{repo_id}:{raw_item["path"]}',
+        item_type=(
+            ItemType.FOLDER if raw_item["type"] in FOLDER_ITEM_TYPES else ItemType.FILE
+        ),
+        item_name=raw_item["name"],
+    )
+
+
+###
+# module-local helpers
+@dataclass(frozen=True, slots=True)
+class Repository(ItemResultable):
+    id: str
+    name: str
+
+    @property
+    def item_result(self) -> ItemResult:
+        return ItemResult(
+            item_id=f"{self.id}:",
+            item_name=self.name,
+            item_type=ItemType.FOLDER,
+        )
```

Comment:
I think you should add here `def from_json()` method to parse **`JSON data`** so It prevents`AttributeError` if there is not such method already, because you are calling this method here `items=[Repository.from_json(item).item_result for item in resp]`. If there is such method just leave It as It is

Reply:
it extends ItemResultable, which inspects dataclass class and makes this automatically